### DISCUSSION TITLE
Improvements for DioException stacktraces in 6.0.0

### DISCUSSION THREAD

**Discussion by kuhnroyal** [cfug/dio#1914] [Ideas]
As this problem came up again in a PR, maybe we can explore some options for dio 6.0.0 here.

Due to async gaps we don't have exactly one stacktrace that contains everything.

We can currently only have one stacktrace for a `DioException` and we previously decided to use the stacktrace that points to the the actual dio call (get/put etc.) instead of the the stacktrace that points deeper into the internals. This stacktrace is more helpful to the user in figuring out which call causes the exception.

Do we want to make both stacktraces accessible?

1. We can wrap the cause of a `DioException` again in some `DioExceptionCause` class and add the original stacktrace there.
2. We can use an `Expando` to attach the original stacktrace without a breaking change and provide a new function `DioException.extractCause(Function(Object? error, StackTrace? stack))` to access it.

**Comment by AlexV525** [comment]
Would the `stack_trace` package help with asynchronous gaps? We can combine all solutions and make stack traces as rich as possible.

**Reply by kuhnroyal** [reply]
We actually had this discussion a while back, but I can't find it. The `stack_trace` package can create `Chain`s around async gaps but at the cost of some overhead due to the use of Zones. Additionally the stack traces are formatted differently.

**Reply by kuhnroyal** [reply]
Found it:
See https://github.com/cfug/dio/pull/1722 & https://github.com/cfug/dio/pull/1723

**Comment by Tienisto** [comment]
The ideal solution would be obviously a native stacktrace solution that traces back to the user call. 

Currently, the developer only sees the useful stacktrace if there is a proper `DioException` handling. The native stacktrace only provides information about internals of Dio:

```dart
import 'package:dio/dio.dart';

void main() async {
  try {
    await Dio().get('https://google.com/abc');
  } catch (e, st) {
    print('Internal stacktrace of dio:');
    print(st);

    if (e is DioException) {
      print('The actual helpful stacktrace:');
      print(e.stackTrace);
    }
  }
}
```

This is not a big deal if the developer depends on dio directly but this does not work for general purpose libraries that do not have a dependency on Dio.

I think the following should be ideal:

```text
#1      Blabla.function (package:some_package/main.dart:111:2)
<asynchronous suspension>
#2      Blabla.function (package:some_package/main.dart:112:4)
<asynchronous suspension>
#3      Blabla.function (package:some_package/main.dart:113:5)
<asynchronous suspension>
```

But we somehow lose the suspension information (maybe due to https://stackoverflow.com/questions/68101983/most-valuable-stacktrace-is-lost-when-using-completer).

**Reply by kuhnroyal** [reply]
This is not possible with the current Dart 2.15.0 constraints, but we should be able to do that if we raise the min Dart constraint to 2.16.0+ which has `Error.throwWithStackTrace()`.
